### PR TITLE
break up selecting keys from decorating in new keys

### PR DESCRIPTION
**Pull request by tcql**
giving functions smaller surface area of responsibility - this gives us more control 💪 

Now, instead of decorateLayer doing key selection + decorating, `selectKeys` is it's own method, with `decorateLayer` handling only the process of applying new keys

old: 

```js
// decorateLayer(layer, keysToKeep, newProps)
Decorator.decorateLayer(layer, ['id', 'type'], [/* ... new properties... */]);
```

new: 
```js
// selectKeys(layer, keysToKeep)
Decorator.selectKeys(layer, ['id', 'type']);
// decorateLayer(layer, newProps)
Decorator.decorateLayer(layer, [/* ... new properties ... */]);
```

This requires implementing clients to do a little extra work but affords more control over what happens and when it happens. For example, in the old setup, it's impossible to decorate a layer with new properties and then specify "keys to keep" on the output side (key selection always applied to the source data. `newProps` was never filtered). It unlocks more fine-grained controls, e.g https://github.com/mapbox/tilelive-decorator/issues/10

This does require an extra loop over `layer.features` to achieve the same work, so I've tested performance indirectly via tilelive-decorator benchmarks;

old: 

```js
decorator x 235 ops/sec ±4.45% (34 runs sampled)
decorator#requiredKeys x 599 ops/sec ±4.86% (35 runs sampled)
```

new: 

```js
decorator x 220 ops/sec ±6.46% (35 runs sampled)
decorator#requiredKeys x 565 ops/sec ±5.31% (35 runs sampled)
```

IMO this is fine, and an even bigger drop would be acceptable to me for the extra 💪  this gets us. I'd like to do a test of this in some real world infrastructure to see how it runs at scale before 🚢 , but I'm happy so far.

### CODE REVIEW

**Comment by tcql** [comment]
real world performance looks good 👍

**Comment by mourner** [comment]
> it's impossible to decorate a layer with new properties and then specify "keys to keep" on the output side (key selection always applied to the source data. newProps was never filtered)

Can you give a few more examples of how this would be used in the real world? E.g. with the example above, would you `decorateLayer` first and then `selectKeys` instead of the other way around?

**Comment by tcql** [comment]
> would you decorateLayer first and then selectKeys instead of the other way around?

@mourner yes exactly. The example in mind is mapbox traffic, which applies records from redis (which may include some extra properties we don't want on the output tile) to a road graph that includes many extra properties. After decorating, we want to specify strictly which properties should exist on the output.

**Comment by mourner** [comment]
@tcql would you achieve the same result by just generating a newProps array with filtered keys first and only then decorating?

**Comment by tcql** [comment]
@mourner hmm yes, that'd be possible. This functionality, however, is more flexible since you don't really need to know where properties came from (new props? or from the source tile?) in order to filter them. 

I liked this solution because specifying keepKeys as part of `decorateLayer` always felt weird to me - adding new properties and filtering what keys to keep are two separate pieces of work, and it's entirely possible you could want to add new keys without specifying which keys to keep at all (just everything in the source + everything that came from redis).

File: test/test.js [hunk 1]
```diff
@@ -118,7 +123,8 @@ test('decorateLayer filtering half of the features', function (t) {
     t.deepEqual(layer.keys, keys);
     var required = ['foo'];
     Decorator.filterByKeys(layer, required);
-    Decorator.decorateLayer(layer, keys, null);
+    Decorator.selectKeys(layer, keys);
+    Decorator.decorateLayer(layer, null);
```
Comment (mourner) [inline test/test.js]: Can you also add a test for `decorateLayer` + `selectKeys` order?

File: index.js [hunk 2]
```diff
@@ -51,43 +52,57 @@ function filterByKeys(layer, requiredKeys) {
...
-        for (var id in newProps[i]) {
-            addKey(id, keyLookup, layer.keys, feature.tags);
-            addValue(newProps[i][id], valLookup, layer.values, feature.tags);
+    }
+
+    return layer;
+}
+
+function decorateLayer(layer, newProps) {
+    if (newProps && (newProps.length !== layer.features.length)) {
```
Comment (mourner) [inline index.js]: You could do `!newProps || (newProps.length !== ...)` and then avoid wrapping the body with the `if` below.

File: index.js [hunk 3]
```diff
@@ -51,43 +52,57 @@ function filterByKeys(layer, requiredKeys) {
...
-        if (!newProps) continue;
-        for (var id in newProps[i]) {
-            addKey(id, keyLookup, layer.keys, feature.tags);
-            addValue(newProps[i][id], valLookup, layer.values, feature.tags);
+    }
+
+    return layer;
+}
+
+function decorateLayer(layer, newProps) {
```
Comment (mourner) [inline index.js]: Since `decorateLayer` is now just updating properties, should we rename it to something like `updateProperties`?

**Review by mourner — CHANGES_REQUESTED** [review]

File: index.js (hunk 3 above)
Comment (tcql) [reply index.js]: mmm, yea... I do like `decorate` or `decorateLayer`, but `updateLayerProperties` is more precise

**Comment by tcql** [comment]
While we're renaming things, we should work out better names for `filterByKeys` and `selectKeys` - currently, the established pattern is for methods to be suffixed with `Layer` when they operate on a single layer as opposed to a tile object: 

- `getLayerValues`
- `decorateLayer`
- `mergeLayer`

so we should probably either follow that pattern or ditch it entirely

**Comment by tcql** [comment]
@mourner updated names 👍

**Review by mourner — APPROVED** [review]

**Pull request merged**